File: svc-position/src/main/java/com/beancounter/position/valuation/Gains.java
package com.beancounter.position.valuation;

import com.beancounter.common.model.MoneyValues;
import java.math.BigDecimal;
import org.springframework.stereotype.Service;

@Service
public class Gains {

  public void value(BigDecimal total, MoneyValues moneyValues) {

    if (!BigDecimal.ZERO.equals(total)) {
      moneyValues.setUnrealisedGain(
          moneyValues.getMarketValue()
              .subtract(moneyValues.getCostValue()));
    }

    moneyValues.setTotalGain(
        moneyValues.getUnrealisedGain()
            .add(moneyValues.getDividends()
                .add(moneyValues.getRealisedGain())));
  }

}

File: svc-position/src/main/java/com/beancounter/position/valuation/Gains.kt
package com.beancounter.position.valuation

import com.beancounter.common.model.MoneyValues
import org.springframework.stereotype.Service
import java.math.BigDecimal

@Service
class Gains {
    fun value(total: BigDecimal, moneyValues: MoneyValues) {
        if (total.compareTo(BigDecimal.ZERO) != 0) {
            moneyValues.unrealisedGain = moneyValues.marketValue
                .subtract(moneyValues.costValue)
        }
        moneyValues.totalGain = moneyValues.unrealisedGain
            .add(
                moneyValues.dividends
                    .add(moneyValues.realisedGain)
            )
    }
}
